pkg/tools/cache.go

```diff
@@ -0,0 +1,147 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/golang/glog"
+	cniv1 "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1"
+	"github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/client/clientset/versioned"
+	"github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/client/informers/externalversions"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+)
+
+type NetAttachDefCache struct {
+	networkAnnotationsMap      map[string]map[string]string
+	networkAnnotationsMapMutex *sync.Mutex
+	stopper                    chan struct{}
+	isRunning                  int32
+}
+
+type NetAttachDefCacheService interface {
+	Start()
+	Stop()
+	Get(namespace string, networkName string) map[string]string
+}
+
+func Create() NetAttachDefCacheService {
+	return &NetAttachDefCache{make(map[string]map[string]string),
+		&sync.Mutex{}, make(chan struct{}), 0}
+}
+
+// Start creates informer for NetworkAttachmentDefinition events and populate the local cache
+func (nc *NetAttachDefCache) Start() {
+	factory := externalversions.NewSharedInformerFactoryWithOptions(setupNetAttachDefClient(), 0, externalversions.WithNamespace(""))
+	informer := factory.K8sCniCncfIo().V1().NetworkAttachmentDefinitions().Informer()
+	// mutex to serialize the events.
+	mutex := &sync.Mutex{}
+
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			mutex.Lock()
+			defer mutex.Unlock()
+			netAttachDef := obj.(*cniv1.NetworkAttachmentDefinition)
+			nc.put(netAttachDef.Namespace, netAttachDef.Name, netAttachDef.Annotations)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			mutex.Lock()
+			defer mutex.Unlock()
+			oldNetAttachDef := oldObj.(*cniv1.NetworkAttachmentDefinition)
+			newNetAttachDef := newObj.(*cniv1.NetworkAttachmentDefinition)
+			if oldNetAttachDef.GetResourceVersion() == newNetAttachDef.GetResourceVersion() {
+				glog.Infof("no change in net-attach-def %s, ignoring update event", nc.getKey(oldNetAttachDef.Namespace, newNetAttachDef.Name))
+				return
+			}
+			nc.remove(oldNetAttachDef.Namespace, oldNetAttachDef.Name)
+			nc.put(newNetAttachDef.Namespace, newNetAttachDef.Name, newNetAttachDef.Annotations)
+		},
+		DeleteFunc: func(obj interface{}) {
+			mutex.Lock()
+			defer mutex.Unlock()
+			netAttachDef := obj.(*cniv1.NetworkAttachmentDefinition)
+			nc.remove(netAttachDef.Namespace, netAttachDef.Name)
+		},
+	})
+	go func() {
+		atomic.StoreInt32(&(nc.isRunning), int32(1))
+		// informer Run blocks until informer is stopped
+		glog.Infof("starting net-attach-def informer")
+		informer.Run(nc.stopper)
```

Comment:
Can we put in a log after .Run "net-attach-def informer stopped"? It may aid debugging.

Reply:
done